src/storage/mvcc/reader/scanner/delta.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! A dedicate scanner that outputs content in each CF.
+
+use engine::CF_DEFAULT;
+
+use crate::storage::kv::SEEK_BOUND;
+use crate::storage::mvcc::write::{Write, WriteType};
+use crate::storage::mvcc::Result;
+use crate::storage::txn::{Result as TxnResult, TxnEntry, TxnEntryScanner};
+use crate::storage::{Cursor, Key, Snapshot, Statistics};
+
+use super::ScannerConfig;
+
+pub struct DeltaScanner<S: Snapshot> {
+    cfg: ScannerConfig<S>,
+    // if is none, we needn't scan it
+    lock_cursor: Option<Cursor<S::Iter>>,
+    write_cursor: Cursor<S::Iter>,
+    default_cursor: Option<Cursor<S::Iter>>,
+    /// Is iteration started
+    is_started: bool,
+    statistics: Statistics,
+    // the end ts is cfg.ts
+    begin_ts: u64,
+}
+
+impl<S: Snapshot> TxnEntryScanner for DeltaScanner<S> {
+    fn next_entry(&mut self) -> TxnResult<Option<TxnEntry>> {
+        Ok(self.read_next()?)
+    }
+    fn take_statistics(&mut self) -> Statistics {
+        std::mem::replace(&mut self.statistics, Statistics::default())
+    }
+}
+
+impl<S: Snapshot> DeltaScanner<S> {
+    pub fn new(
+        cfg: ScannerConfig<S>,
+        lock_cursor: Option<Cursor<S::Iter>>,
+        write_cursor: Cursor<S::Iter>,
+        begin_ts: u64,
+    ) -> DeltaScanner<S> {
+        DeltaScanner {
+            cfg,
+            lock_cursor,
+            write_cursor,
+            statistics: Statistics::default(),
+            default_cursor: None,
+            is_started: false,
+            begin_ts,
+        }
+    }
+    pub fn read_next(&mut self) -> Result<Option<TxnEntry>> {
+        self.check_and_start()?;
+        if self.lock_cursor.is_none() {
+            return self.read_next_write_key();
+        }
+        unimplemented!()
+    }
+
+    pub fn read_next_write_key(&mut self) -> Result<Option<TxnEntry>> {
+        loop {
+            if !self.write_cursor.valid()? {
+                return Ok(None);
+            }
+            let w_key = self.write_cursor.key(&mut self.statistics.write);
+            let (user_key, commit_ts) = Key::split_on_ts_for(w_key)?;
+            // Use `from_encoded_slice` to reserve space for ts, so later we can append ts to
+            // the key or its clones without reallocation.
+            let current_user_key = Key::from_encoded_slice(user_key);
+            // goto next key
+            if commit_ts < self.begin_ts {
```

Comment:
We want it to read ts range (begin_ts, end_ts], so I think it should be 
```suggestion
            if commit_ts <= self.begin_ts {
```